clients/cobol-lsp-vscode-extension/src/services/ProcessorGroups.ts

```diff
@@ -206,50 +215,105 @@ function pathMatches(program: string, documentPath: string) {
   }
   const cfgPath = path.join(ws[0], PROCESSOR_GROUP_FOLDER);
   const procCfgPath = path.join(cfgPath, PROCESSOR_GROUP_PROC);
+  if (!fs.existsSync(procCfgPath)) {
+    return undefined;
+  }
+  return fs.readFileSync(procCfgPath).toString();
+}
+
+export const readProgramConfigFileContent = (): string | undefined => {
+  const ws = SettingsUtils.getWorkspaceFoldersPath(true);
+  if (ws.length < 1) {
+    return undefined;
+  }
+  const cfgPath = path.join(ws[0], PROCESSOR_GROUP_FOLDER);
   const pgmCfgPath = path.join(cfgPath, PROCESSOR_GROUP_PGM);
-  if (!fs.existsSync(pgmCfgPath) || !fs.existsSync(procCfgPath)) {
+  if (!fs.existsSync(pgmCfgPath)) {
     return undefined;
   }
-  const procCfg: ProcessorsConfig = JSON.parse(
-    fs.readFileSync(procCfgPath).toString(),
-  );
-  const pgmCfg: ProgramsConfig = JSON.parse(
-    fs.readFileSync(pgmCfgPath).toString(),
-  );
-  const pgroup = matchProcessorGroup(pgmCfg, documentPath, ws[0]);
+  return fs.readFileSync(pgmCfgPath).toString();
+};
+
+export const loadProcessorsConfigForDocument = (
+  documentUri: string,
+  processorGroupsContent: string | undefined,
+  programConfigContent: string | undefined,
+  b4gContent: string | undefined,
+): ProcessorConfig | undefined => {
+  const documentPath = Uri.parse(documentUri).fsPath;
+  const ws = SettingsUtils.getWorkspaceFoldersPath(true);
+  if (ws.length < 1) {
+    return undefined;
+  }
+  if (processorGroupsContent === undefined) {
+    return undefined;
+  }
+  const pgroups = loadProcessorConfigurations(processorGroupsContent);
+  const pgmCfg: ProgramsConfig =
+    programConfigContent === undefined
+      ? { pgms: [] }
+      : JSON.parse(programConfigContent);
+
+  const b4g = b4gContent ? decodeBridgeJson(b4gContent) : undefined;
+  const pgroup = selectProcessorGroup(pgmCfg, documentPath, ws[0], b4g);
 
   let result;
-  procCfg.pgroups.forEach((p) => {
+  pgroups.forEach((p) => {
     if (pgroup === p.name) {
       result = p;
       return;
     }
   });
   return result;
+};
+
+function selectProcessorGroup(
+  pgmCfg: ProgramsConfig,
+  documentPath: string,
+  workspacePath: string,
+  b4g: B4GTypeMetadata | undefined,
+): string | undefined {
+  if (b4g === undefined) {
+    return matchProcessorGroup(pgmCfg, documentPath, workspacePath);
+  }
+  const selectedElement = b4g.fileExtension
+    ? path.basename(documentPath, "." + b4g.fileExtension)
+    : path.basename(documentPath);
+  return b4g.elements[selectedElement] === undefined
+    ? b4g.defaultProcessorGroup
+    : b4g.elements[selectedElement].processorGroup;
 }
 
 function loadProcessorGroupSettings<T>(
-  scopeUri: string,
+  documentUri: string,
   atrtibute: string,
   configObject: T,
   dialect: string = "COBOL",
 ): T {
+  const pgCfg: ProcessorConfig | undefined = loadProcessorsConfigForDocument(
+    documentUri,
+    readProcessorGroupsFileContent(),
+    readProgramConfigFileContent(),
+    loadBridgeJsonContent(Uri.parse(documentUri).fsPath),
```

Comment:
One more suggestion - maybe all these functions should return the already parsed JSON, or undefined when failures occur whether due to reading or parsing? 
It looks like everything is parsed upfront in `loadProcessorsConfigForDocument` at the moment anyway.

Reply:
make sense